Ruby rebuild + cmake => 3.29.2

satmandu opened the pull request:
- Rebuilds `ruby` from `ruby_3_3` branch to fix https://github.com/advisories/GHSA-592j-995h-p23j
- Add `crewlog(string)` function to `crew`.
- Refactor json package removal for upgraded packages.
- Add `save_json` function to safely save the json file.
- Automatically rerun `crew upgrade` when `ruby` is being upgraded.
- Added `CREW_ESSENTIAL_PACKAGES = %w[gcc_lib glibc gmp ruby zlibpkg zstd]` to `const.rb`
- If `distcc` is installed, use `distcc -j` for `CREW_NPROC` calculation.
- grep usage in `commands/remove.rb` needs quoting for file names with spaces.

Works properly:
- [x] `x86_64`
- [x] `i686`
- [x] `armv7l` <!-- (reasons why it doesn't) -->

### Run the following to get this pull request's changes locally for testing.
```
CREW_REPO=https://github.com/satmandu/chromebrew.git CREW_BRANCH=ruby_cmake crew update ; yes | crew upgrade -v
```

<!--
## That's it
Thank you for submitting your pull request.
When done, please delete the parts of this template which you don't need or these, which are only for guidance.
-->

satmandu commented:
Currently there's a problem where `CREW_REPO=https://github.com/satmandu/chromebrew.git CREW_BRANCH=ruby_cmake crew update ; yes | crew upgrade` doesn't work:

```bash
chronos@rpi52-armv7l /usr/local/lib/crew/packages $ CREW_REPO=https://github.com/satmandu/chromebrew.git CREW_BRANCH=ruby_cmake crew update ; yes | crew upgrade
remote: Enumerating objects: 43, done.
remote: Counting objects: 100% (43/43), done.
remote: Compressing objects: 100% (4/4), done.
remote: Total 26 (delta 22), reused 26 (delta 22), pack-reused 0
Unpacking objects: 100% (26/26), 15.37 KiB | 215.00 KiB/s, done.
From https://github.com/satmandu/chromebrew
 * branch                ruby_cmake -> FETCH_HEAD
HEAD is now at 111cbde75 cmake => 3.29.2
Package lists, crew, and library updated.
Checking for package updates...
ruby could be updated from 3.3.0 to 3.3.0-2f65458
cmake could be updated from 3.29.0 to 3.29.2
                                                                                                                                                                                                              2 packages can be updated.
Run `crew upgrade` to update all packages or `crew upgrade <package1> [<package2> ...]` to update specific packages.
Updating packages...
ruby: Ruby is a dynamic, open source programming language with a focus on simplicity and productivity.
Performing pre-flight checks...
Precompiled binary available, downloading...
Cannot find cached archive. 😔 Will download.
Ruby archive downloaded.
Unpacking archive using 'tar', this may take a while...
Removing since upgrade or reinstall...
Refusing to remove essential package ruby.
Performing pre-install...
Performing install...
Using rdfind to convert duplicate files to hard links.
Now scanning ".", found 17612 files.
Now have 17612 files in total.
Removed 15 files due to nonunique device and inode.
Total size is 40856842 bytes or 39 MiB
Removed 1997 files due to unique sizes from list. 15600 files left.
Now eliminating candidates based on first bytes: removed 15481 files from list. 119 files left.
Now eliminating candidates based on last bytes: removed 68 files from list. 51 files left.
Now eliminating candidates based on sha1 checksum: removed 51 files from list. 0 files left.
It seems like you have 0 files that are not unique
Totally, 0 B can be reduced.
Now making hard links.
Making 0 links.
Performing post-install for ruby...
Updating ruby gems. This may take a while...
Ruby installed!
Ruby was upgraded. Please run 'crew upgrade' again to make sure upgrades are complete.
chronos@rpi52-armv7l /usr/local/lib/crew/packages $ crew upgrade
Updating packages...
ruby: Ruby is a dynamic, open source programming language with a focus on simplicity and productivity.
Performing pre-flight checks...
Precompiled binary available, downloading...
Archive found in cache
Unpacking archive using 'tar', this may take a while...
Removing since upgrade or reinstall...
Refusing to remove essential package ruby.
Performing pre-install...
Performing install...
Using rdfind to convert duplicate files to hard links.
Now scanning ".", found 17612 files.
Now have 17612 files in total.
Removed 15 files due to nonunique device and inode.
Total size is 40856842 bytes or 39 MiB
Removed 1997 files due to unique sizes from list. 15600 files left.
Now eliminating candidates based on first bytes: removed 15481 files from list. 119 files left.
Now eliminating candidates based on last bytes: removed 68 files from list. 51 files left.
Now eliminating candidates based on sha1 checksum: removed 51 files from list. 0 files left.
It seems like you have 0 files that are not unique
Totally, 0 B can be reduced.
Now making hard links.
Making 0 links.
Performing post-install for ruby...
Updating ruby gems. This may take a while...
Ruby installed!
Ruby was upgraded. Please run 'crew upgrade' again to make sure upgrades are complete.
chronos@rpi52-armv7l /usr/local/lib/crew/packages $
```

uberhacker commented on packages/ruby.rb:
```diff
@@ -3,44 +3,43 @@
-  depends_on 'libyaml' # This is needed to install gems
+  depends_on 'zlibpkg' # R
+
+  conflicts_ok
```
What is conflicting?  Is it really OK?  If there is another ruby package that has conflicting files, removing that package could brick crew?

satmandu replied:
Only conflicts with Ruby ...

satmandu replied:
Conflict code is wonky ... 🤦‍♂️😔

Zopolis4 commented on packages/cmake.rb:
```diff
@@ -3,23 +3,24 @@
   depends_on 'bzip2' => :build
   depends_on 'cppdap' # R
   depends_on 'curl' # R
+  depends_on 'distcc' => :build
```
Does this really need to be the case?

satmandu replied:
Just trying to speed up the build. Tests take forever... Using distcc helps...

satmandu commented:
Added some debug code which shows this happening:
```bash
Removing since upgrade or reinstall...
Refusing to remove essential package ruby.
Performing pre-install...
Performing install...
Using rdfind to convert duplicate files to hard links.
Now scanning ".", found 17612 files.
Now have 17612 files in total.
Removed 15 files due to nonunique device and inode.
Total size is 43133583 bytes or 41 MiB
Removed 2002 files due to unique sizes from list. 15595 files left.
Now eliminating candidates based on first bytes: removed 15478 files from list. 117 files left.
Now eliminating candidates based on last bytes: removed 68 files from list. 49 files left.
Now eliminating candidates based on sha1 checksum: removed 49 files from list. 0 files left.
It seems like you have 0 files that are not unique
Totally, 0 B can be reduced.
Now making hard links.
Making 0 links.
before json change
{
  "name": "ruby",
  "version": "3.3.0",
  "binary_sha256": "a97a17336b4d233129fb5a55c82d32b38b981d017cda70ba3cbb51012bc129ca"
}
{
  "name": "ruby",
  "version": "3.3.0-2f65458",
  "sha256": "c554a429edbc1d104be2f7a149f91b3892e81605dc550f26a918c2f8bb749c22"
}
after json change
{
  "name": "ruby",
  "version": "3.3.0",
  "binary_sha256": "a97a17336b4d233129fb5a55c82d32b38b981d017cda70ba3cbb51012bc129ca"
}
{
  "name": "ruby",
  "version": "3.3.0-2f65458",
  "sha256": "c554a429edbc1d104be2f7a149f91b3892e81605dc550f26a918c2f8bb749c22"
}
{
  "name": "ruby",
  "version": "3.3.0-2f65458",
  "sha256": "c554a429edbc1d104be2f7a149f91b3892e81605dc550f26a918c2f8bb749c22"
}
Performing post-install for ruby...
```

Zopolis4 commented:
Is this happening for any other package?

uberhacker commented on bin/crew:
```diff
@@ -1283,15 +1283,18 @@ def install
   system "/usr/bin/jq --arg key ruby -e '.installed_packages[] | select(.name == \$key )' /usr/local/etc/crew/device.json"
+
+  # Clean out package from json before adding, especially if package may be in CREW_ESSENTIAL_PACKAGES.
+  @device[:installed_packages].delete_if { |entry| entry['name'] == @pkg.name }
```
Why is this necessary?  Doesn't it just get added back in again?  Wouldn't a better approach be to check if the package exists and if not, add it?

satmandu replied:
A second one gets added in if it isn't first removed.

We need
`If the name is in the array
  remove it
then add name
...
`

uberhacker replied:
A second one is added only because you're not checking if it exists beforehand.

satmandu replied:
Adjusted code to remove the entry if it is in `CREW_ESSENTIAL_PACKAGES`.

satmandu commented:
Everything seems to work now.

uberhacker commented on bin/crew:
```diff
@@ -115,15 +115,34 @@ at_exit do
   # load_json(): (re)load device.json
   json_path = File.join(CREW_CONFIG_PATH, 'device.json')
   @device   = JSON.load_file(json_path, symbolize_names: true)
+  @device_json = JSON.load_file(json_path)
```
I sure would like to keep this simple.  Why load twice?

satmandu replied:
Because there is different behavior whether `symbolize_names: true` is set or not. That's the prior behavior, and @Zopolis4 has been doing stuff that works without it being set, so I kept both until we figure out if we actually do need `symbolize_names: true` set everywhere...

Zopolis4 replied:
We don't need it set at all-- it's just a question of the way we access values (i.e. :name vs 'name')

uberhacker replied:
What is the default for symbolize_names?  If we don't need to set it, let's not set it.  I'd be more inclined to want it to be flexible for symbols or names if that is possible.

Zopolis4 replied:
It is not set by default (neither is our custom method to symbolize values), and no method for supporting both is available. 

Feel free to change it and all the code that relies on it, or I'll eventually refactor all of the code.

satmandu replied:
Looks like just flipping it to not set it breaks things:
```
chronos@cheekon-x86_64 /usr/local/lib/crew/packages $ crew upgrade
/usr/local/bin/crew:394:in `upgrade': undefined method `each' for nil (NoMethodError)

    @device[:installed_packages].each do |pkg|
                                ^^^^^
        from /usr/local/bin/crew:1878:in `upgrade_command'
        from /usr/local/bin/crew:1908:in `<main>'
chronos@cheekon-x86_64 /usr/local/lib/crew/packages $
```

satmandu replied:
uh... well, maybe I can fix that...

Zopolis4 replied:
by and large you'll have to replace :foo with 'foo' whenever json is concerned.

satmandu replied:
not easily... I think this should be left for a cleanup in a future PR...

satmandu replied:
@Zopolis4 Going in reverse, code like `@device[:installed_packages].delete_if { |entry| entry['name'] == @pkg.name }` doesn't work. Any idea how to fix that? (That's only used twice, so it might be easier to take care of those...)

satmandu replied:
I think I got it working...

satmandu replied:
Ok I removed it...

uberhacker commented on bin/crew:
```diff
@@ -1250,12 +1274,23 @@ def install
+
+    if CREW_ESSENTIAL_PACKAGES.include?(@pkg.name)
+      crewlog "Removing package #{@pkg.name} from device.json"
+      @device_json['installed_packages'].delete_if { |entry| entry['name'] == @pkg.name }
```
Still don't quite understand why the package needs to be removed from device.json.

satmandu replied:
Upgrade is the union of remove and install.

Remove takes out the package's line from the json file, except when one of the `CREW_ESSENTIAL_PACKAGES` is being dealt with, in which case remove returns early and the json file never gets updated.

This works around that, as otherwise, a new line gets added to the json file
This is a problem because the version detection for packages looks for the first package info returned from the json file, and since the old info is still there after an upgrade due to this bug, crew thinks that the package still needs to be upgraded after the new package is installed.

Zopolis4 replied:
Removing the package should remove the package from the skin file, though.

Zopolis4 replied:
Ah, I see-- the removal is getting skipped, so the package isn't getting removed.

uberhacker commented on bin/crew:
```diff
@@ -1274,14 +1309,8 @@ def install
-  # Copy over original if the write to the tmp file succeeds.
-  FileUtils.cp "#{CREW_CONFIG_PATH}/device.json.tmp", File.join(CREW_CONFIG_PATH, 'device.json')
-  FileUtils.rm "#{CREW_CONFIG_PATH}/device.json.tmp"
+  @device_json['installed_packages'].push(name: @pkg.name, version: @pkg.version, sha256: PackageUtils.get_sha256(@pkg, build_from_source: @opt_source))
```
Check if package is already in installed_packages prior to push?

satmandu replied:
If the package is already installed we still need to delete the package info... We could do a check for that, if that would make more sense?

satmandu replied:
I've added a package installed check right before it is added.  (Right now we DO have a situation where the package keeps getting added...)

uberhacker commented on lib/const.rb:
```diff
@@ -115,8 +117,12 @@
+               ENV.fetch('CREW_NPROC', `distcc -j`.chomp)
+             else
+               ENV.fetch('CREW_NPROC', `nproc`.chomp)
+             end
```
What's the difference?

satmandu replied:
With multiple distcc hosts, the number can be much higher than what nproc gives, as work can be offloaded to other machines...

uberhacker commented on packages/cmake.rb:
```diff
@@ -44,6 +44,9 @@ class Cmake < CMake
   # CMakeLib.testDebuggerNamedPipe-Script (armv7l,i686,x86_64)
   # RunCMake.CMakeRelease (armv7l,i686,x86_64)
   def self.check
+    @current_installed_cmake = `cmake --version | head -n 1 | awk '{print \$3}'`.chomp
```
What happens if cmake is not installed?

uberhacker commented on packages/ruby.rb:
```diff
@@ -3,44 +3,41 @@
   description 'Ruby is a dynamic, open source programming language with a focus on simplicity and productivity.'
   homepage 'https://www.ruby-lang.org/en/'
-  version '3.3.0' # Do not use @_ver here, it will break the installer.
+  version '3.3.0-2f65458' # Do not use @_ver here, it will break the installer.
```
Does it make sense to wait until the next stable release?  I only say this because the maintainers probably have a good reason for not tagging a new release.

Zopolis4 replied:
This is for a security fix, so it might make sense to get this through before a new stable.

satmandu replied:
Also, this is from their stable branch... the update will trigger a gem rebuild for any vulnerable versions installed.

uberhacker reviewed (commented):
See comments.

Zopolis4 commented:
How does upgrading an essential package work then, given that we don't (to my knowledge) remove the currently running ruby binary?

satmandu commented:
> How does upgrading an essential package work then, given that we don't (to my knowledge) remove the currently running ruby binary?

Crew should overwrite the package without removing it, thus never hopefully creating a moment when there aren't working binaries for the package on disk. (Which isn't entirely true, especially if doing a massive number of updates and dependencies get mucked up during the update... but that's arguably an edge case.)

uberhacker commented on bin/crew:
```diff
@@ -421,7 +439,12 @@ def upgrade(*pkgs, build_from_source: false)
+    at_exit do
+      puts "Ruby was upgraded. Rerunning 'crew upgrade' to make sure upgrades are complete.".lightblue
+      exec 'crew upgrade'
+    end
```
Can we move line 441 to this conditional so it makes more logical sense?
```
else
  puts 'Packages have been updated.'.lightgreen
end
```

satmandu replied:
Done.
